src/select/SelectMenu.ts

```diff
@@ -1,77 +1,54 @@
 /* eslint-disable react-hooks/exhaustive-deps */
 import React from "react";
-import { createOnKeyDown } from "reakit-utils";
-import { BoxHTMLProps } from "reakit/ts/Box/Box";
-import { SelectStateReturn } from "./useSelectState";
-import { createHook, createComponent } from "reakit-system";
+import { usePopover, PopoverHTMLProps } from "reakit";
+import { SelectStateReturn } from "./SelectState";
 import { SELECT_KEYS } from "./__keys";
-
-export type SelectMenuOptions = Pick<
-  SelectStateReturn,
-  "setTypehead" | "selected" | "hide" | "show" | "visible"
-> & {
-  onChange?: (value: any) => void;
-};
-
-const useSelectMenu = createHook<SelectMenuOptions, BoxHTMLProps>({
+import { createComponent, createHook } from "reakit-system";
+import {
+  useComposite,
+  CompositeProps,
+  CompositeHTMLProps,
+} from "reakit/Composite";
+import { usePortalShortcut } from "./__utils";
+import { useForkRef } from "reakit-utils";
+
+export type SelectMenuOptions = CompositeProps &
+  Pick<
+    SelectStateReturn,
+    "setSelected" | "visible" | "values" | "currentId" | "move"
+  > & {
+    modal?: boolean;
+  };
+
+export type SelectMenuHTMLProps = CompositeHTMLProps & PopoverHTMLProps;
+
+const useSelectMenu = createHook<SelectMenuOptions, SelectMenuHTMLProps>({
   name: "SelectMenu",
+  compose: [useComposite, usePopover],
   keys: SELECT_KEYS,
+  useOptions({ modal = false, ...options }) {
+    return { modal, ...options };
+  },
 
   useProps(
-    { setTypehead, selected, onChange, hide, show, visible },
-    { ...htmlProps },
+    { visible, move, values, currentId },
+    { ref: htmlRef, style: htmlStyle, ...htmlProps },
   ) {
-    const keyClear = React.useRef<any>(null);
-    const [typed, setTyped] = React.useState("");
-
-    React.useEffect(() => {
-      onChange && onChange(selected);
-    }, [selected]);
-
-    const onKeyDown = React.useMemo(() => {
-      return createOnKeyDown({
-        stopPropagation: true,
-        keyMap: () => ({
-          Escape: hide,
-          ArrowUp: show,
-          ArrowDown: show,
-        }),
-      });
-    }, []);
+    const ref = React.useRef<HTMLElement>(null);
 
-    const clearKeyStrokes = () => {
-      setTypehead(typed);
-
-      if (keyClear.current) {
-        clearTimeout(keyClear.current);
-        keyClear.current = null;
-      }
-
-      keyClear.current = setTimeout(() => {
-        setTyped("");
-        keyClear.current = null;
-      }, 800);
-    };
+    usePortalShortcut(ref, { values, currentId, move });
 
     React.useEffect(() => {
-      if (typed !== "") {
-        clearKeyStrokes();
+      if (values) {
```

Comment:
```suggestion
    if (values && !selected?.length) {
```

Do this additional check so that the select doesn't focus the first option, when the value is selected

Reply:
I don't know what do you meant by this but adding this check prevented the original behaviour

Reply:
When it is multiselect, it can focus the first option, itself. 🤔

Reply:
I think we can do this check

```js
if (values && !selected?.length) {
        const firstItem = values[0];
        firstItem && move?.(firstItem.id);
      } else {
        const item = values.find(i => {
          return i.value === selected[selected.length - 1];
        })?.id;
        move?.(item as string);
      }
```

On Multiselect the focus would start from the last selected item

Reply:
Great !!!